[3.x] Add maximum roughness cutoff to SSR to improve performance

Calinou opened the pull request:
`3.x` version of https://github.com/godotengine/godot/pull/56804.

In a test scene with mixed rough and non-rough materials, this saves 0.2 ms of GPU time with very little visual artifacting (GeForce RTX 4090, 3840×2160). This also incorporates the roughness fadeout from https://github.com/godotengine/godot/pull/69828, which can help hide visual artifacts on rough surfaces and generally make SSR look more stable.

~~The roughness cutoff can be adjusted in the Environment properties. Setting it to `1.0` will result in the previous behavior being used.~~
**Edit (June 2023):** This is no longer present for now. We need to discuss whether we should add this setting, as it's not in 4.x and probably won't be added there.

**Testing project (`3.x`):** [test_ssr_performance-3.x.zip](https://github.com/godotengine/godot/files/7875340/test_ssr_performance-3.x.zip)

## Preview

### No roughness cutoff (previous behavior)

![before webp](https://github.com/godotengine/godot/assets/180032/633104f6-e8c7-4de3-ae5f-b0f083ec6c57)

### Roughness cutoff = 0.7 (new default)

![after webp](https://github.com/godotengine/godot/assets/180032/748260a2-16ec-4e11-8c18-a7a7adaee003)

Calinou committed:
Add maximum roughness cutoff to SSR to improve performance

In a test scene with mixed rough and non-rough materials, this saves
0.15-0.3 ms of GPU time with very little visual artifacting
(GTX 1080, 2560×1440).

The roughness cutoff can be adjusted in the Environment properties.

clayjohn commented on drivers/gles3/shaders/screen_space_reflection.glsl:
```diff
@@ -62,6 +63,12 @@ void main() {
 
 	float roughness = normal_roughness.w;
 
+	if (roughness > max_roughness) {
+		// Do not compute SSR for rough materials to improve performance at the cost of
+		// subtle artifacting.
+		return;
+	}
+
```
```suggestion
	if (roughness > max_roughness) {
		// Do not compute SSR for rough materials to improve performance at the cost of
		// subtle artifacting.
		frag_color = vec4(0.0);
		return;
	}

```

Without writing to frag_color an early return can result in uninitialized memory leaking out.

clayjohn reviewed (commented):
I'm not sure if adding this to 3.x makes sense. This will add a sharp cutoff point where SSR will just disappear. In 4.0 the equivalent code was added after we added a smooth falloff to the SSR rough reflections so that SSR could gradually fade out on materials with a varying amount of roughness. A similar fade-out could be added here and then I guess it would be fine

Calinou commented:
> A similar fade-out could be added here and then I guess it would be fine

Yes, I agree backporting both the fade-out and roughness cutoff makes sense. SSR looking too sharp and too present in 3.x is a common complaint after all.

Calinou commented:
Rebased and tested again, it works as expected. I've implemented @clayjohn's roughness cutoff from 4.x. However, this poses a challenge for the `max_roughness` setting, as the fadeout values don't depend on the roughness cutoff. I've tried `min()`/`max()`ing those with the `max_roughness` values, but it tends to look pretty bad in practice if you change the roughness cutoff from the default.

Screenshots in OP were updated to show the new visual and performance results.
